From e2423f62513defcc3b5b370a955b3866cb7c10f1 Mon Sep 17 00:00:00 2001
From: Georg Hackel <georg.hackel@9elements.com>
Date: Wed, 19 Aug 2026 16:36:25 +0200
Subject: [PATCH] dt-bindings: usb: add HPE USB Device Controllers binding

Add device tree binding documentation for the HPE GXP USB Device controller (aka gadget).

The udcg goes to soc as it's just internal glue.

Upstream-Status: Pending
Assisted-by: Claude:claude-sonnet-5
Signed-off-by: Georg Hackel <georg.hackel@9elements.com>
---
 .../bindings/soc/hpe/hpe,gxp-udcg.yaml        | 39 +++++++++++
 .../devicetree/bindings/usb/hpe,gxp-udc.yaml  | 69 +++++++++++++++++++
 2 files changed, 108 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/soc/hpe/hpe,gxp-udcg.yaml
 create mode 100644 Documentation/devicetree/bindings/usb/hpe,gxp-udc.yaml

diff --git a/Documentation/devicetree/bindings/soc/hpe/hpe,gxp-udcg.yaml b/Documentation/devicetree/bindings/soc/hpe/hpe,gxp-udcg.yaml
new file mode 100644
index 000000000000..8f37d4696a7a
--- /dev/null
+++ b/Documentation/devicetree/bindings/soc/hpe/hpe,gxp-udcg.yaml
@@ -0,0 +1,39 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/soc/hpe/hpe,gxp-udcg.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: HPE GXP UDCG (USB Device Controller group)
+
+maintainers:
+  - Marcello Sylvester Bauer <marcello.bauer@9elements.com>
+  - Tan Siewert <tan.siewert@9elements.com>
+
+description: |
+  System configuration registers for the HPE GXP USB Device Controller
+  (UDC) glue logic. This region holds control/status bits shared across
+  the UDC instance(s) and is accessed as a syscon by the UDC driver via
+  the hpe,udcg phandle.
+
+properties:
+  compatible:
+    items:
+      - const: hpe,gxp-udcg
+      - const: syscon
+
+  reg:
+    maxItems: 1
+
+required:
+  - compatible
+  - reg
+
+additionalProperties: false
+
+examples:
+  - |
+    udcg: syscon@80400800 {
+        compatible = "hpe,gxp-udcg", "syscon";
+        reg = <0x80400800 0x200>;
+    };
diff --git a/Documentation/devicetree/bindings/usb/hpe,gxp-udc.yaml b/Documentation/devicetree/bindings/usb/hpe,gxp-udc.yaml
new file mode 100644
index 000000000000..0114d13b5b3e
--- /dev/null
+++ b/Documentation/devicetree/bindings/usb/hpe,gxp-udc.yaml
@@ -0,0 +1,69 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/usb/hpe,gxp-udc.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: HPE GXP UDC (USB Device Controller)
+
+maintainers:
+  - Marcello Sylvester Bauer <marcello.bauer@9elements.com>
+  - Tan Siewert <tan.siewert@9elements.com>
+
+description: |
+  The HPE GXP SoC includes up to 8 virtual USB device controllers that share a
+  common group controller (UDCG). Each virtual device supports up to 7 flexible
+  endpoints and operates at USB High Speed. It is used to present virtual USB
+  devices to a connected host.
+  The group controller is defined at the soc level as hpe,gxp-udcg.
+
+properties:
+  compatible:
+    const: hpe,gxp-udc
+
+  reg:
+    maxItems: 1
+
+  interrupts:
+    maxItems: 1
+
+
+  hpe,vdev-num:
+    $ref: /schemas/types.yaml#/definitions/uint32
+    description:
+      Virtual device instance number, used to select which internal
+      UDC channel/context this node controls on SoCs with multiple
+      UDC instances.
+
+  hpe,num-flex-endpoints:
+    $ref: /schemas/types.yaml#/definitions/uint32
+    description:
+      Number of flexible (configurable) USB endpoints implemented by
+      this UDC instance, in addition to the fixed control endpoint.
+
+  hpe,udcg:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description:
+      Phandle to the syscon node containing shared UDC glue/configuration
+      registers.
+
+required:
+  - compatible
+  - reg
+  - interrupts
+  - hpe,vdev-num
+  - hpe,udcg
+
+additionalProperties: false
+
+examples:
+  - |
+    udc0: usb@80401000 {
+      compatible = "hpe,gxp-udc";
+      reg = <0x80401000 0x1000>;
+      interrupts = <13>;
+      interrupt-parent = <&vic1>;
+      hpe,vdev-num = <0>;
+      hpe,num-flex-endpoints = <4>;
+      hpe,udcg = <&udcg>;
+    };
